lib/puppet/gettext/config.rb

```diff
@@ -30,12 +47,43 @@ def self.translation_mode(conf_path)
     end
   end
 
+  # Prevent future gettext initializations
+  def self.disable_gettext
+    @gettext_disabled = true
+  end
+
   # Attempt to initialize the gettext-setup gem
   # @param path [String] to gettext config file
   # @param file_format [Symbol] translation file format to use, either :po or :mo
   # @return true if initialization succeeded, false otherwise
   def self.initialize(conf_file_location, file_format)
-    # Bypass gettext until we can resolve a performance regression related to it, PUP-8009.
-    return false
+    return if @gettext_disabled
+
+    unless file_format == :po || file_format == :mo
+      raise Puppet::Error, "Unsupported translation file format #{file_format}; please use :po or :mo"
+    end
+
+    return false unless @gettext_found
+
+    if conf_file_location && File.exists?(conf_file_location)
+      if GettextSetup.method(:initialize).parameters.count == 1
+        # For use with old gettext-setup gem versions, will load PO files only
+        GettextSetup.initialize(conf_file_location)
+      else
+        GettextSetup.initialize(conf_file_location, :file_format => file_format)
+      end
+      # Only change this once.
```

Comment:
Doesn't this end up getting called every time `lib/puppet/module.rb:431` calls `if Puppet::GettextConfig.initialize(locales_path, :po)`?

Reply:
I wonder about this too. Maybe we should separate it out so we only call it from puppet.rb.

Reply:
This is intentional, to fix https://tickets.puppetlabs.com/browse/PUP-7839

Reply:
I'm not sure I understand... I'm referring to the `Only change this once` part that appears to be calling `FastGettext.locale = GettextSetup.negotiate_locale(Locale.current.language)` over and over again.

I would think once we've set `FastGettext.locale`, we want to leave it set, no?

Reply:
Oh yeah, could FastGettext.locale always be default_locale?

Reply:
That's true. In the English case, this will get called for each module (but only once, since if we got to this point, https://github.com/puppetlabs/puppet/blob/master/lib/puppet/module.rb#L427 will trigger next time). If we think this is causing a performance hit, the tradeoff is that if, for example, a person has a German computer and translates their module into German, those translations will never be loaded unless Puppet itself also has German translations.

Reply:
And I guess the reason the comment says "only _change_ this once" is because if no translations are present, the negotiated locale equals the default locale, so the value of FastGettext.locale doesn't change.

Reply:
Environment caching seems like a reasonable solution for this.

Reply:
How so?

Reply:
We won't create a new Module instance as often with environment caching turned on.